Moving GC for badger

vyzo opened the pull request:
Cherry-picked from #6728.

This adds the option to perform moving GC on a badger blockstore; this effectively reclaims all space.
It is also hooked into the splitstore, to perform moving GC once every 20 compactions (about once a week, user configurable).

Rationale: badger supports online GC, which is quite fast but ultimately ineffective at reclaiming space. It has been observed that the hotstore size slowly creeps up, despite frequent garbage collections.
This addresses the situation with real moving GC.

vyzo commented:
dunno what's up with the gen-check failure, there is no information there.

magik6k commented:
> dunno what's up with the gen-check failure, there is no information there.

When you expand the second-to-last thing on circle it shows you the diff. The usual solution is to just run `make gen` which should regenerate whatever it's complaining about (in this case config doc-gen stuff which just landed in https://github.com/filecoin-project/lotus/pull/6848 - so just running `make cfgdoc-gen` should also be enough, and faster)

magik6k commented on node/config/types.go:
```diff
@@ -282,7 +282,8 @@ type Splitstore struct {
 
-	HotStoreMessageRetention uint64
+	HotStoreMessageRetention  uint64
+	HotStoreMovingGCFrequency uint64
```
Would be nice to add docstrings to those since we now expose those comments in user configs automagically

vyzo replied:
Ah nice, I didn't know about this. I will add a one-liner for them, as we have detailed documentation in the README.

vyzo replied:
done.

vyzo commented:
Effectiveness of moving GC:

Before moving GC, after about a week with online GC in every compaction:
```
132G    ./datastore/splitstore/hot.badger
```

An hour after moving GC:
```
57G     ./datastore/splitstore/hot.badger.1627141269
```

vyzo commented:
rebased on master.

vyzo commented:
rebased on master and fixed conflicts.

raulk commented on blockstore/badger/blockstore.go:
```diff
@@ -80,13 +83,26 @@ const (
 	stateClosed
 )
 
+const (
+	moveStateNone = iota
+	moveStateMoving
+	moveStateCleanup
+	moveStateLock
+)
```
1. let's make this an int type
2. godocs on what each of these states means?

vyzo replied:
sure!

vyzo replied:
done.

raulk commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+	opts.Dir = path
+	opts.ValueDir = path
+
+	db2, err := badger.Open(opts.Options)
+	if err != nil {
+		return fmt.Errorf("failed to open badger blockstore in %s: %w", path, err)
+	}
```
What would happen if we fail to open a new badger store?

vyzo replied:
it would just error out and signal an error in the `CollectGarbage` call -- I don't think we need to do anything more here.

raulk commented on blockstore/badger/blockstore.go:
```diff
@@ -349,18 +698,36 @@ func (b *Blockstore) Put(block blocks.Block) error {
+		return err
+	}
+
+	if b.db2 != nil {
+		if err := put(b.db2); err != nil {
+			return err
+		}
+	}
```
What is this for?

vyzo replied:
so that we don't lose data during the move -- the put (also putMany) has to go to both dbs during the move.

raulk commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+
+// doCopy copies a badger blockstore to another, with an optional filter; if the filter
+// is not nil, then only cids that satisfy the filter will be copied.
+func (b *Blockstore) doCopy(from, to *badger.DB, filter func(cid.Cid) bool) error {
```
What do we plan to use this filter for?

vyzo replied:
The filter is there to support a `CopyTo` interface which would take a filter as we've been discussing with steven.
We can also remove it for now and add it later, when we have a concrete use case for it.

vyzo replied:
removed it.

raulk commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+	defer putPooled()
+
+	var buf []byte
+	for iter.Rewind(); iter.Valid(); iter.Next() {
```
Did we consider using https://pkg.go.dev/github.com/dgraph-io/badger/v2#readme-stream, which is more efficient and automatically concurrent, instead of an iterator? It also supports selecting keys by predicates (to satisfy the filter).

vyzo replied:
My only concern with the streaming interface is that we also have writes happening concurrently with the move.
If that's not a problem, then by all means, let's do streaming, I would be very happy to go faster :)

vyzo replied:
Ok, I checked the docs and there is no mention of concurrent writes issues; I would assume that this is fine.
I'll try to convert to use the streaming interface.

vyzo replied:
Turns out the code is way simpler with the streaming interface; only concern is allocations, as it will make a hole bunch of intermediate garbage with the protobuf objects.

raulk replied:
I'd be surprised if they don't use object pooling. Maybe check their source? Otherwise Lotus is lucky to have a memory watchdog...

vyzo replied:
yeah, good point, We'll see how well it works in practice, but I expect it to be quite a bit faster.

vyzo replied:
bingo! It's 4.5 times faster (4.5min -vs- 9min) and didn't notice any deleterious effects from memory usage.

raulk reviewed (commented):
Quite a lot to digest -- tricky code. Still reviewing.

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+		if filter != nil {
+			k := kk
+			if b.prefixing {
+				k = k[b.prefixLen:]
```
I'd just check the length first, just to be safe.

vyzo replied:
ok.

vyzo replied:
filter is gone.

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+func (b *Blockstore) deleteDB(path string) {
+	// follow symbolic links, otherwise the data wil be left behind
+	lpath := path
+	for {
```
EvalSymlinks?

vyzo replied:
aha! I didn't know this method existed; I'll use it.

vyzo replied:
done.

vyzo commented on node/config/types.go:
```diff
@@ -278,11 +278,23 @@ type Chainstore struct {
+	// HotStoreMessageRetention specifies the retention policy for messages, in finalities beyond
+	// the compaction boundary; default is 0.
 	HotStoreMessageRetention uint64
+	// HotStoreMovingGCFrequency specifies how often to perform moving GC on the hotstore.
```
sure.

vyzo replied:
done.

vyzo commented on blockstore/blockstore.go:
```diff
@@ -37,9 +37,21 @@ type BlockstoreIterator interface {
+
+// BlockstoreMovingGCPath is a garbage collection option that specifies an optional
+// target path for moving GC.
+var BlockstoreMovingGCPath = blockstoreMovingGCPath{}
```
Well the use-case is a symlinked badger in a separate file system, where the user specifies the path.

Stebalien replied:
By _telling_ badger where to move the blockstore on GC? I'd expect that to be a _static_ config option (or maybe the moving GC would follow the symlink on start).

Let's talk at standup, I'm confused.

vyzo replied:
Yeah, I'll drop it for now and if/when we need it we can re-add it, and we can talk on standup.

vyzo replied:
done.

vyzo commented on blockstore/blockstore.go:
```diff
@@ -37,9 +37,21 @@ type BlockstoreIterator interface {
 // BlockstoreGC is a trait for blockstores that support online garbage collection
 type BlockstoreGC interface {
-	CollectGarbage() error
+	CollectGarbage(options map[interface{}]interface{}) error
```
The difficulty is that the functional option would be defined in the base type and used in a subtype.
I guess we could make a `GCConfig` object and apply there so that we avoid the parsing.
So yeah, I'll change it to use functional options, it's cleaner.

Stebalien replied:
Basically, we have two cases:

1. This doesn't generalize. Why are we pretending it does?
2. It does generalize. So generalize it.

One-off abstractions are just a pain.

vyzo replied:
yup, I respect your strong feelings on the matter!

vyzo replied:
done, functional options it is!

vyzo commented on blockstore/blockstore.go:
```diff
@@ -37,9 +37,21 @@ type BlockstoreIterator interface {
+
+// BlockstoreMovingGC is a garbage collection option that instructs the blockstore
+// to use moving GC if supported.
+var BlockstoreMovingGC = blockstoreMovingGCKey{}
```
Good point, I'll rename to `FullGC` or something similar.

vyzo commented on blockstore/splitstore/README.md:
```diff
@@ -59,6 +59,17 @@ These are options in the `[Chainstore.Splitstore]` section of the configuration:
   nodes beyond 4 finalities, while running with the discard coldstore option.
   It is also useful for miners who accept deals and need to lookback messages beyond
   the 4 finalities, which would otherwise hit the coldstore.
+- `HotStoreMovingGCFrequency` -- specifies how frequenty to garbage collect the hotstore
```
yup, will rename to `FullGCFrequency`.

raulk replied:
IMO "moving" is already the most pessimistic/common-denominator form of GC. I do think it generalises nicely because it's the ultimate form of "GC" that databases that do not support efficient GC will fall back to. It's basically a backstop/catch-all that is applicable to all and any database.

vyzo replied:
That's a good point, but I think that steven's "FullGC" nomenclature is right -- the user shouldn't care about the mechanics, as long as it really is a full gc; besides I have already renamed it :)
(but I can rename back if we deem so)

vyzo replied:
done.

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -35,6 +36,8 @@ var (
 	log = logger.Logger("badgerbs")
 )
 
+const moveBatchSize = 16384
```
sure.

vyzo replied:
it's gone!

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+	ok := true
+	err = os.Rename(dbpath, oldpath)
+	if err != nil {
+		// this is bad, but not catastrophic; new data will be written in db2 and user can fix
```
then we should panic I guess.

Stebalien replied:
Yeah... that may be the only real option. Although I'd prefer to handle the error better (at least maybe retry this operation once?).

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -80,13 +83,26 @@ const (
+	rlock     int
+
+	db   *badger.DB
+	db2  *badger.DB // when moving
```
hrm, maybe, just more characters to type :)

Stebalien replied:
Yeah, just slightly more descriptive. Also, https://github.com/joaotavora/eglot or https://github.com/emacs-lsp/lsp-mode.

vyzo replied:
done.

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+		err = os.Symlink(path, dbpath)
+		if err != nil {
+			// ditto, this is bad, but not catastrophic; user can fix
+			log.Errorf("error symlinking badger db dir from %s to %s; USER ACTION REQUIRED", path, dbpath)
```
It's not catastrophic in the sense that we haven't lost any data.
It is bad however, since as you point out the db is not in the target path because the symlinking failed.
I guess we should panic to make the user fix the situation instead of leaving him with a ticking bomb.

Stebalien replied:
Yeah. I guess, not catastrophic but also not "can continue".

vyzo replied:
done, it now panics.

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+	iter := txn.NewIterator(opts)
+	defer iter.Close()
+
+	pooled := make([][]byte, 0, 2*moveBatchSize)
```
Ah yes, good idea.

vyzo replied:
I implemented it, and then I did the streaming interface; turns out it's completely unnecessary.

vyzo commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+		b.deleteDB(oldpath)
+	}
+
+	log.Info("moving blockstore done")
```
at this point we have swapped the blockstores in the internal db variables and using the new one.
I guess we should panic in the two bad conditions to not even get here.

Stebalien replied:
Yeah, you're right.

vyzo replied:
done.

vyzo commented:
Changes per review comments (quite a few!):
- Updated the interface to use functional options.
- Renamed MovingGC to FullGC in the nomenclature.
- Removed the target path option and made the new path resolve links.
- Dropped the filter from the copy; we are not currently using it anyway.
- Implemented the copy using the streaming interface.
- Made path manipulation failures that require action panic; that seems to be the only sane way to proceed.
- Made the state variables typed.
- Renamed db2 to dbNext in the data structure.
- Used EvalSymlinks where appropriate

vyzo commented:
ok, this is ready for the next round of review; in the meantime I'll deploy it in my discard node to get some timings and real-world burn.

vyzo commented:
Timings are looking really good:
```
2021-07-27T11:31:33.039Z        INFO    splitstore      splitstore/splitstore_gc.go:30  garbage collecting hotstore done        {"took": 535.387156281}
```

previously it was taking about 40min; that's 4 times faster.

jennijuju commented:
We really want this in, however, it's not a hard blocker for v1.11.1-rc1

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+	if basePath == linkPath {
+		path = basePath
+	} else {
+		name := filepath.Base(basePath)
+		dir := filepath.Dir(linkPath)
+		path = filepath.Join(dir, name)
```
Why are we doing this instead of just using `linkPath`? Assuming `/a/b/foo -> /x/y/bar`, this will create `/x/y/foo.$time`. Don't we want to create `/x/y/bar.$time`?

vyzo replied:
Well, the problem with that is that when we have multiple moves we'll have `foo.$time` already and we'll end up linking in `foo.$time.$time....`.

vyzo replied:
I'll add a comment here to make the intent clear.

Stebalien replied:
I see. I was just thinking, if the user linked to a directory on a remote drive...

vyzo replied:
added comment about intent in https://github.com/filecoin-project/lotus/pull/6905/commits/950ded454d2a6101090097f4e3bb60c722875497

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+	b.moveCond.Broadcast()
+	b.moveMx.Unlock()
+
+	var path string
```
`newPath`? Just calling this path is confusing.

vyzo replied:
sure, will fix the names to make the logic easier to follow.

vyzo replied:
done in https://github.com/filecoin-project/lotus/pull/6905/commits/950ded454d2a6101090097f4e3bb60c722875497

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+	b.dbNext = nil
+	b.unlockMove(moveStateCleanup)
+
+	err = db1.Close()
```
dbOld, please! All these 1's and 2's make the code very difficult to follow.

vyzo replied:
yup! sorry about that.

Stebalien replied:
Thank you! (sorry to be a bit anal about this)

vyzo replied:
not a problem, really. We will want to read this code again in the future, and I can see how the 1s and 2s could confuse pretty much everyone (including me :)

vyzo replied:
done in https://github.com/filecoin-project/lotus/pull/6905/commits/950ded454d2a6101090097f4e3bb60c722875497

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+		log.Warnf("error closing old badger db: %s", err)
+	}
+
+	dbpath := b.opts.Dir
```
_which_ db path is this? Please chose a variable name.

Stebalien replied:
As far as i can tell, this is the _old_ path.

vyzo replied:
this is the canonical path where we want the db to live.

Stebalien replied:
Ah, got it. Leaving this as it is sounds reasonable then.

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+	}
+
+	dbpath := b.opts.Dir
+	oldpath := fmt.Sprintf("%s.old.%d", dbpath, time.Now().Unix())
```
Backup path?

vyzo replied:
sure.

vyzo replied:
done in https://github.com/filecoin-project/lotus/pull/6905/commits/950ded454d2a6101090097f4e3bb60c722875497

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+		panic(fmt.Errorf("error renaming old badger db dir from %s to %s: %w; USER ACTION REQUIRED", dbpath, oldpath, err)) //nolint
+	}
+
+	if err = os.Symlink(path, dbpath); err != nil {
```
What if I wasn't using symlinks?

vyzo replied:
is that a problem? 
The issue with not using symlinks is moving across filesystems, which can take a long time.

Stebalien replied:
Basically, what if I had the database at "foo". Now I'll end up with a symlink at "foo" pointing to a random directory.

If I started out with plain directories, I'd rather end up with plain directories. It's not critical, but the current version is going to confuse users.

1. They initialize their repo.
2. They see where their hotstore is.
3. The hotstore gets moved and replaced by a symlink.

vyzo replied:
Hrm, it complicates things to try and keep it symlink-free, as we'll need two versions of things.

We should probably be clear to users that the splitstore _owns_ the splitstore directory; they shouldn't muck with it.
The coldstore is another matter though.

Stebalien replied:
Fair enough.

Stebalien reviewed (commented):
Mostly naming comments but also some questions around file names.

TL;DR: When I have 2 databases and 4 different paths (old, new, link target, backup), the names really matter.

Stebalien commented on blockstore/badger/blockstore.go:
```diff
@@ -165,26 +194,239 @@ func (b *Blockstore) isOpen() bool {
+	dbpath := b.opts.Dir
+	oldpath := fmt.Sprintf("%s.old.%d", dbpath, time.Now().Unix())
+
+	if err = os.Rename(dbpath, oldpath); err != nil {
```
Won't this just rename the symlink? Don't we need to move the actual file?

We need integration tests for this.

Stebalien replied:
Ah, I see... we are renaming the symlink intentionally. Why not just delete it then delete `linkPath`?

vyzo replied:
yeah, it's fine -- we'll delete deep in deleteDB which follows symlinks.

Also, I don't think we need an integration test for this, we can test pretty much everything with unit tests as the changes are pretty self-contained.

What do you want to add to the unit test?

Stebalien replied:
> yeah, it's fine -- we'll delete deep in deleteDB which follows symlinks.

It's fine, but a _bit_ weird.

> What do you want to add to the unit test?

The unit test is fine. I thought there was a bug here which would have been caught by a basic test. I revisited this because I then _did_ look at the test, and it clearly checked this case.

Your tests are good.

Stebalien commented:
Other feedback: The renaming logic with the symlink dance is strange. We should be preserving symlinks if the user is using them, and _not_ using symlinks if the user isn't already using them.

1. Not all filesystems support them.
2. Are we sure we're not creating absolute symlinks? That could cause trouble when moving repos.

vyzo commented:
> Not all filesystems support them.

Wait, is there a realistic situation where lotus would run in a filesystem that doesn't support symlinks?
Even windows support them these days.

> Are we sure we're not creating absolute symlinks? That could cause trouble when moving repos.

Gah, you are right about this. We _are_ creating absolute links, I'll fix in follow up.

Stebalien commented:
> Wait, is there a realistic situation where lotus would run in a filesystem that doesn't support symlinks?

Well... probably not. Given that NFS does. Although someone probably does it anyways...

Sorry, I've had some recent encounters with symlinks.

vyzo commented:
Follw up in #6905
